linkerd/tls/src/server.rs

```diff
@@ -17,19 +19,39 @@ use tokio_rustls::server::TlsStream;
 use tower::util::ServiceExt;
 use tracing::{debug, trace, warn};
 
-pub trait HasConfig {
-    fn tls_server_name(&self) -> identity::Name;
-    fn tls_server_config(&self) -> Arc<Config>;
-}
+pub type Config = Arc<rustls::ServerConfig>;
 
 /// Produces a server config that fails to handshake all connections.
-pub fn empty_config() -> Arc<Config> {
+pub fn empty_config() -> Config {
     let verifier = rustls::NoClientAuth::new();
-    Arc::new(Config::new(verifier))
+    Arc::new(rustls::ServerConfig::new(verifier))
+}
+
+/// A marker type for remote client idenities..
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct ClientId(pub id::Name);
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum NoTls {
+    /// Identity is administratively disabled.
+    Disabled,
+
+    /// No TLS is wanted because the connection is a loopback connection which
+    /// doesn't need or support TLS.
+    Loopback,
+
+    /// No TLS is wanted because the connection is a loopback connection which
+    /// doesn't need or support TLS.
+    PortSkipped,
+
+    // TLS not established by the remote client.
+    NoTlsFromRemote,
 }
 
+pub type ConditionalTls = Conditional<Option<ClientId>, NoTls>;
```

Comment:
nit/TIOLI: it's a little bit weird that the corresponding type in the client module is `ConditionalServerId`, but this is `ConditionalTls`...but, it makes sense because here, on the accept side, we may be doing TLS without a client identity if it's not mutual TLS. I think the names should stay as they are, but it's maybe worth a comment explaining this, though?